Support generating field metadata for some runtime introspection

zackyancey opened the pull request:
resolves #79 

Here's what it looks like:

```rs
#[bitfield(u32, introspect)]
struct Bitfield {
    #[bits(0..=7)]
    a: u8,
    #[bits(8..=15)]
    b: [u8; 2],
    #[bits([24..=25, 30..=31])]
    c: u4,
    #[bit(29)]
    d: bool,
}

// <NAME>_BITS exposes the "bits" value(s):
assert_eq!(Bitfield::A_BITS, 0..=7);
assert_eq!(Bitfield::B_BITS, 8..=15);
assert_eq!(Bitfield::C_BITS, [24..=25, 30..=31]);
assert_eq!(Bitfield::D_BITS, 29..=29);

// Arrays also have <NAME>_COUNT and <NAME>_STRIDE
assert_eq!(Bitfield::B_COUNT, 2);
assert_eq!(Bitfield::B_STRIDE, 8);

// <name>_mask() returns a mask for the field
assert_eq!(Bitfield::a_mask(), 0x000000FF);
assert_eq!(Bitfield::b_mask(0), 0x0000FF00);
assert_eq!(Bitfield::b_mask(1), 0x00FF0000);
assert_eq!(Bitfield::c_mask(), 0xC3000000);
assert_eq!(Bitfield::d_mask(), 0x20000000);
```

A couple design decisions that may be worth discussing:

* `*_BITS` is a single range for contiguous bitfields, and a slice of ranges for non-contiguous ones. It could be simpler to always generate a slice (eg. `A_BITS == [0..=7]`), but since most fields are contiguous and the single range is much more ergonomic in that case, I went with that. I'm a bit on the fence about that decision though.
* `*_mask()` is a const method, so that we can accomodate arrays. The main alternatives would be:
  * Have a single mask constant, and the user can shift based on `*_COUNT` and `*_STRIDE`. That would work, but since we've already generated 90% of the work, may as well give them a function.
  * Generate mask as a slice (eg `const B_MASK = [0xFF00, [0xFF0000]`). That could be a lot of constants for large arrays though.
* These values can be enabled per-struct with an attribute, or globally with a feature. Having it as a feature allows users to turn it on by default, and also to get introspection data from external defintions that may not have included the attribute.
* I'm open to changing the `introspect` name, as I think introspection implies a bit more than what we're generating here. Not really sure what else to go with though.

danlehmann commented:
This looks really great. I have a minor suggestion for adding a test, but otherwise I'd like to take this as-is

zackyancey commented:
Added a changelog entry. What's your test suggestion? The one test I added to `bitfield_tests.rs` is pretty basic, so if there are any edge cases you think are worth adding, I can add them.

zackyancey commented on bitbybit/src/bitfield/mod.rs:
```diff
@@ -359,6 +372,40 @@ fn setter_name(field_name: &Ident) -> Ident {
+    // The field might have started with r#. If so, it was likely used for a keyword. This can be dropped here
+    let field_name_without_prefix = {
+        let s = field_name.to_string();
+        if let Some(s) = s.strip_prefix("r#") {
```
Good call, added both.

danlehmann commented:
Thanks for your contribution!
